dali/kernels/shape.h

```diff
@@ -0,0 +1,609 @@
+  reverse_iterator rend() noexcept { return shape.rend(); }
+  const_reverse_iterator rbegin() const noexcept { return shape.rbegin(); }
+  const_reverse_iterator rend() const noexcept { return shape.rend(); }
+  const_reverse_iterator crbegin() const noexcept { return shape.crbegin(); }
+  const_reverse_iterator crend() const noexcept { return shape.crend(); }
+
+  size_type size() const { return shape.size(); }
+
+  Container shape;
+  static constexpr int static_ndim = ndim;
+
+  template <int other_ndim>
+  TensorShape<other_ndim> first();
+  template <int other_ndim>
+  TensorShape<other_ndim> last();
+
+  TensorShape<DynamicDimensions> first(int count);
+  TensorShape<DynamicDimensions> last(int count);
+
+ protected:
+  // Disallow instatiation of Base class
+  TensorShapeBase() : shape{} {}
+
+  TensorShapeBase(const Container &c) : shape(c) {}
+  TensorShapeBase(Container &&c) : shape(std::move(c)) {}
+};
+
+template <>
+struct TensorShape<DynamicDimensions>
+    : public TensorShapeBase<std::vector<int64_t>, DynamicDimensions> {
+  using Base = TensorShapeBase<std::vector<int64_t>, DynamicDimensions>;
+  TensorShape(const std::vector<int64_t> &s) : Base(s) {}
+
+  template <size_t N>
+  TensorShape(std::array<int64_t, N> s) : Base(typename Base::container_type(s.begin(), s.end())) {}
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : Base(typename Base::container_type{i0, int64_t{s}...}) {}
+
+  TensorShape() = default;
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <int other_ndim>
+  TensorShape(const TensorShape<other_ndim> &other)
+      : Base(typename Base::container_type(other.shape.begin(), other.shape.end())) {}
+
+  template <int other_ndim>
+  TensorShape &operator=(const TensorShape<other_ndim> &other) {
+    shape = Base::container_type(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  // We allow only explicit conversion from dynamic to static dim
+  template <int other_ndim>
+  TensorShape<other_ndim> to_static() const {
+    static_assert(other_ndim != DynamicDimensions,
+                  "Conversion to static only allowed for static shape");
+    assert(size() == other_ndim);
+    TensorShape<other_ndim> shape;
+    for (int i = 0; i < other_ndim; i++) {
+      shape[i] = (*this)[i];
+    }
+    return shape;
+  }
+
+  void resize(typename Base::size_type count) { shape.resize(count); }
+
+  void resize(typename Base::size_type count, const typename Base::value_type &val) {
+    shape.resize(count, val);
+  }
+};
+
+template <int ndim>
+struct TensorShape : public TensorShapeBase<std::array<int64_t, ndim>, ndim> {
+  using Base = TensorShapeBase<std::array<int64_t, ndim>, ndim>;
+  TensorShape(const std::array<int64_t, ndim> &s) : Base(s) {}
+  // We just want zero-initialization
```

Comment:
This comment does say what we want, not what we do. Either remove it or refer to the base class doing the right thing (which it does).

Reply:
Done